packages/firestore/src/api/database.ts

```diff
@@ -1936,13 +1968,40 @@ export class Query implements firestore.Query {
             firstOrderByField
           );
         }
-      } else if (filter.op === RelationOp.ARRAY_CONTAINS) {
-        if (this._query.hasArrayContainsFilter()) {
-          throw new FirestoreError(
-            Code.INVALID_ARGUMENT,
-            'Invalid query. Queries only support a single array-contains ' +
-              'filter.'
-          );
+      } else if (
+        [
+          RelationOp.ARRAY_CONTAINS,
+          RelationOp.ARRAY_CONTAINS_ANY,
+          RelationOp.IN
+        ].indexOf(filter.op) >= 0
+      ) {
+        // Check if any of the above 3 filters are already present in the query.
+        const existingFilter =
+          this._query.getRelationOpFilter(RelationOp.ARRAY_CONTAINS) ||
+          this._query.getRelationOpFilter(RelationOp.ARRAY_CONTAINS_ANY) ||
+          this._query.getRelationOpFilter(RelationOp.IN);
+        const hasInAndArrayContains =
+          (existingFilter === RelationOp.ARRAY_CONTAINS &&
+            filter.op === RelationOp.IN) ||
+          (existingFilter === RelationOp.IN &&
+            filter.op === RelationOp.ARRAY_CONTAINS);
+        // The only permitted combination of these 3 filters is IN with ARRAY_CONTAINS.
+        if (hasInAndArrayContains) {
```

Comment:
I think the following filter will incorrectly be allowed by this code: `.where('foo', 'array_contains', 5).where('bar', 'in', [1, 2, 3]).where('baz', 'in', [4, 5, 6])`

The problem is that existingFilter will be the _first_ of the matching filter types, but there could be more than one (since you can combine in / array-contains)